text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,454 @@
+inconsistent with each other.  The `sampled` flag is equivalent to the
+expression `p - 1 <= r`.  When the invariant `sampled <=> p - 1 <= r`
+is violated, the `ParentBased` sampler MUST correct the propagated
+values as discussed below.
+
+The violation is always addressed by honoring the `sampled` flag and
+setting `log_head_adjusted_count` to either 0 (Unknown) or 63 (Zero).
+
+If `sampled` is false and the invariant is bilated, drop `p` from the
+outgoing context to convey unknown head probability.  Set
+`log_head_adjusted_count` to 0.
+
+The case where `sampled` is true with `p=63` indicating 0% probability
+may by regarded as a special case to allow zero adjusted count
+sampling, which permits non-probabilistic sampling to take place in
+the presence of probability sampling.  Set `log_head_adjusted_count`
+to 63.
+
+If `sampled` is true with `p<63`, drop `p` from the outgoing context
+to convey unknown head probability.  Set `log_head_adjusted_count` to
+0.
+
+## Prototype
+
+[This proposal has been prototyped in the OTel-Go
+SDK.](https://github.com/open-telemetry/opentelemetry-go/pull/2177) No
+changes in the OTel-Go Tracing SDK's `Sampler` or `tracestate` APIs
+were needed.
+
+## Trade-offs and mitigations
+
+### Naming question
+
+This proposal changes the logic of the `TraceIDRatioBased` sampler,
+currently part of the OpenTelemetry specification, in a way that makes
+the name no longer meaningful.  The proposed sampler may be named
+`ConsistentSampler` and the existing `TraceIDRatioBased` sampler can
+be deprecated.
+
+Many SDKs already implement the `TraceIDRatioBased` sampler and it has
+been used for probability sampling at trace roots with arbitrary
+(i.e., not power-of-two) probabilities.  Because of this, we may keep
+the current (under-specified) `TraceIDRatioBased` sampler and rename
+it `ProbabilitySampler` to convey that it does behave in a specified
+way with respect to the bits of the TraceID.
+
+### Not using TraceID randomness
+
+It would be possible, if TraceID were specified to have at least 61
+uniform random bits, to compute the randomness value described above
+as the number of leading zeros among those 61 random bits.
+
+However, this would require modifying the W3C traceparent specification,
+therefore we do not propose to use bits of the TraceID.
+
+[This issue has been filed with the W3C trace context group.](https://github.com/w3c/trace-context/issues/463)
+
+### Not using TraceID hashing
+
+It would be possible to make a consistent sampling decision by hashing
+the TraceID, but we feel such an approach is not sufficient for making
+unbiased sampling decisions.  It is seen as a relatively difficult
+task to define and specify a good enough hashing function, much less
+to have it implemented in multiple languages.
+
+Hashing is also computationally expensive. This proposal uses extra
+data to avoid the computational cost of hashing TraceIDs.
```

Comment:
I would disagree with this argument, it is more expensive to parse the strings from the tracestate, communicate them over network, etc. than a hash of a 16-byte array (at least this is my intuition).

Reply:
@oertl has defended this point in the past with a link to https://github.com/rurban/smhasher, which shows how expensive and failure-prone a number of hashing functions are. My position is that the math of computing adjusted counts from inverse probabilities is likely to have artifacts as a result of low-quality hashing. The better long-term solution is a W3C `traceparent` with 62 specified uniform random bits.